### PR TITLE
Explore: Allow supplementary query data provider to be undefined for mixed data sources

### LINKED ISSUE #65360: Logs Samples: Panel not shown as required in mixed datasources
Follow up from https://github.com/grafana/grafana/issues/65041
https://github.com/grafana/grafana/pull/65357 fixes the bug for all datasources except of mixed datasources. 

With recent changes to mixed data source the behavior of setting the data provider in the state (which is `Obervable<DataQueryResponse>`) changed):

Before: it could be `Obervable<DataQueryResponse> | undefined`
After (only for mixed ds): It's always  `Observable<DataQueryResponse>`. The reason why it's not returning `undefined` is that the logic to get the data provider from a data source is synchronous, but mixed data source needs to get list of data sources asynchronously. The asynchronous logic to get data sources was put inside the observable.

The main drawback of the above is that a provider is always returned. The only way to check if such provider will provide any data is to subscribe to it and check if the response was `{ state: NotStarted }`.

This, however is a problem for LogsSample which requires knowing if supplementary query data will be provided without subscribing it, beacuse LogsSample can be turned on/off on demand. Before (and in non-mixed) this can be simply done by checking if a data provider exists (if it's `undefined` it means the provided target will no emit any supplementary data).

### PR DESCRIPTION
**Pull request by ifrost**
This PR brings back the behavior that supplementary query data provider can be set as `undefined`. Previously, mixed data source would always create an observable making it impossible to check if it's available synchronously (required in https://github.com/grafana/grafana/blob/main/public/app/features/explore/Explore.tsx#L577)

The fix is to pull up the logic to get list of data sources before action to set the provider is called. It means setting the provider will not happen synchronously inside `runQueries`, and may be delayed a bit, but at least it will be set to `undefined` if the query does not support supplementary queries.

<details><summary>before</summary>

Data provider is always returned, data sources are resolved asynchronously inside the provider logic.

```mermaid
sequenceDiagram
participant E as explore/query
participant G as getSupplementaryQueryProvider
participant O as Observable<DataQueryResponse>
participant D as DataSourceSrv


E ->> G : getSupplementaryQueryProvider()
G ->> O : always create a provider
O -->> G : dataProvider: Obervable<DataQueryResponse>
G -->> E : dataProvider: Obervable<DataQueryResponse>
E ->> O : subscribe
activate O
loop query
  O ->> D : await get(query.dataSource)
  D -->> O : ds: DataSourceApi
  alt (!ds.getProvider)
    O --> D : { data: [], state: NotStarted }
  end
end 
O -->> E : { data: [], state: NotStarted }
note right of E : data provider always returns<br/>data, when query doesn't support<br/>supplementary queries it emits<br/>{ data: [], state: NotStarted }
deactivate O
```
</details>

<details><summary>after</summary>

Data sources are resolved first, and the it's determined if data provider is available:

```mermaid
sequenceDiagram
participant E as explore/query
participant G as getSupplementaryQueryProvider
participant O as Observable<DataQueryResponse>
participant D as DataSourceSrv

loop query
  E ->> D : await get(query.dataSource)
  D -->> E : ds: DataSourceApi
  E -->> E : datasources.push(ds)
end 

E ->> G : getSupplementaryQueryProvider(datasources)
opt at least one provider exists
    G ->> O : 
    O -->> G : dataProvider: Obervable<DataQueryResponse>
    G -->> E : dataProvider: Obervable<DataQueryResponse>
    note right of E : data provider is returned<br/>only when provider query emits data
end
opt no providers
    G -->> E : undefined
end
```

</details>

Fixes #65360

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/explore/utils/supplementaryQueries.test.ts [hunk 1]
```diff
@@ -145,17 +144,24 @@ jest.mock('@grafana/runtime', () => ({
...
 
-  return getSupplementaryQueryProvider(rootDataSourceApiMock, type, requestMock, explorePanelDataMock);
+  for (const i in targetSources) {
+    const source = targetSources[i];
+    const datasource = await getDataSourceSrv().get({ uid: source });
+    datasources.push({
+      datasource,
+      targets: [new MockQuery(`${i}`, 'a', { uid: source })],
+    });
+  }
```
Comment (Elfo404) [inline public/app/features/explore/utils/supplementaryQueries.test.ts:162]: you should be able to do this concurrently with something like:

```ts
const datasources = await Promise.all(
    targetSources.map(async (source, i) => {
      const datasource = await getDataSourceSrv().get({ uid: source });
      return {
        datasource,
        targets: [new MockQuery(`${i}`, 'a', { uid: source })],
      };
    })
  );
```

File: public/app/features/explore/state/query.ts [hunk 2]
```diff
@@ -607,51 +615,115 @@ export const runQueries = (
...
+  newQuerySource: Observable<ExplorePanelData>;
+  supplementaryQueries: SupplementaryQueries;
+  queries: DataQuery[];
+  absoluteRange: AbsoluteTimeRange;
+};
+
+const groupDataQueries = async (queries: DataQuery[], scopedVars: ScopedVars) => {
+  queries = queries.filter((t) => {
+    return t.datasource?.uid !== MIXED_DATASOURCE_NAME;
+  });
```
Comment (Elfo404) [inline public/app/features/explore/state/query.ts:649]: not a big fan of reassigning parameters, what about:

```suggestion
  const nonMixedQueries = queries.filter((t) => {
    return t.datasource?.uid !== MIXED_DATASOURCE_NAME;
  });
```

File: public/app/features/explore/state/query.ts [hunk 3]
```diff
@@ -607,51 +615,115 @@ export const runQueries = (
...
+
+  for (const key in sets) {
+    const targets = sets[key];
+    mixed.push({
+      datasource: await getDataSourceSrv().get(targets[0].datasource, scopedVars),
+      targets,
+    });
+  }
+
+  return mixed;
```
Comment (Elfo404) [inline public/app/features/explore/state/query.ts:662]: you can run those concurrently:

```suggestion
  return await Promise.all(
    Object.values(sets).map(async (targets) => {
      const datasource = await getDataSourceSrv().get(targets[0].datasource, scopedVars);
      return {
        datasource,
        targets,
      };
    })
  );
```

File: public/app/features/explore/state/query.ts [hunk 4]
```diff
@@ -607,51 +615,115 @@ export const runQueries = (
...
 
+type HandleSupplementaryQueriesOptions = {
+  exploreId: ExploreId;
+  transaction: QueryTransaction;
+  datasourceInstance: DataSourceApi;
+  newQuerySource: Observable<ExplorePanelData>;
+  supplementaryQueries: SupplementaryQueries;
+  queries: DataQuery[];
+  absoluteRange: AbsoluteTimeRange;
+};
```
Comment (Elfo404) [inline public/app/features/explore/state/query.ts:644]: can we move this close to `handleSupplementaryQueries`?

File: public/app/features/explore/state/query.ts [hunk 5]
```diff
@@ -607,51 +615,115 @@ export const runQueries = (
...
+      absoluteRange,
+    }: HandleSupplementaryQueriesOptions,
+    { dispatch }
+  ) => {
+    let datasources;
+    if (datasourceInstance.meta.mixed) {
+      datasources = await groupDataQueries(transaction.request.targets, transaction.request.scopedVars);
+    } else {
+      datasources = [{ datasource: datasourceInstance, targets: transaction.request.targets }];
+    }
```
Comment (Elfo404) [inline public/app/features/explore/state/query.ts:684]: can we use a more accurate name for the `datasources` variable?
something like `queriesByDatasource` or `groupedQueries` maybe?

File: public/app/features/explore/utils/supplementaryQueries.ts [hunk 6]
```diff
@@ -129,98 +127,66 @@ const getSupplementaryQueryFallback = (
...
-      scan<DataQueryResponse, DataQueryResponse>(
-        (acc, next) => {
-          if (acc.error || next.state === LoadingState.NotStarted) {
-            return acc;
-          }
+  const definedProviders: Array<Observable<DataQueryResponse>> = providers.filter(
+    (item: Observable<DataQueryResponse> | undefined): item is Observable<DataQueryResponse> => {
+      return !!item;
+    }
+  );
```
Comment (Elfo404) [inline public/app/features/explore/utils/supplementaryQueries.ts:156]: you can simply use `isTruthy` from `@grafana/data`, it should yield the same result/types:

```suggestion
  const definedProviders = providers.filter(isTruthy);
```

File: public/app/features/explore/utils/supplementaryQueries.ts [hunk 7]
```diff
@@ -129,98 +127,66 @@ const getSupplementaryQueryFallback = (
...
+    return definedProviders[0];
+  }
 
-          if (next.state && next.state !== LoadingState.Done) {
-            return acc;
-          }
+  return merge(...definedProviders).pipe(
+    scan<DataQueryResponse, DataQueryResponse>(
+      (acc, next) => {
+        if (acc.error || next.state === LoadingState.NotStarted) {
```
Comment (Elfo404) [inline public/app/features/explore/utils/supplementaryQueries.ts:167]: `error` is deprecated, should we also check for `errors`?

**Review by Elfo404 — CHANGES_REQUESTED** [review]

File: public/app/features/explore/utils/supplementaryQueries.test.ts (hunk 1 above)
Comment (ifrost) [reply public/app/features/explore/utils/supplementaryQueries.test.ts]: Fixed in c6339c147e017355ae093e2d4ba059de62608f24

File: public/app/features/explore/state/query.ts (hunk 2 above)
Comment (ifrost) [reply public/app/features/explore/state/query.ts]: Updated in 4ca1ad33d143c963e685bc6d91a38ac99b331375.

File: public/app/features/explore/state/query.ts (hunk 4 above)
Comment (ifrost) [reply public/app/features/explore/state/query.ts:661]: Changed in 5d09a5368737be2f793029277c2cbd172ce546ab.

File: public/app/features/explore/state/query.ts (hunk 5 above)
Comment (ifrost) [reply public/app/features/explore/state/query.ts]: Updated in 5bd2458a1f921a13e432599af119655747cd6120.

File: public/app/features/explore/utils/supplementaryQueries.ts (hunk 6 above)
Comment (ifrost) [reply public/app/features/explore/utils/supplementaryQueries.ts]: Sweet! Exactly what I was looking for. Changed in c21cee1cb0add4cb177c56f4c41dd2d7551d3203.

File: public/app/features/explore/utils/supplementaryQueries.ts (hunk 7 above)
Comment (ifrost) [reply public/app/features/explore/utils/supplementaryQueries.ts]: Switched to `errors` in 2f5a09453a865a359266b3ed9bfea8eca7f43c48.

**Review by Elfo404 — APPROVED** [review]
seems to work as expected, LGTM

**Comment by ifrost** [comment]
Backport: https://github.com/grafana/grafana/pull/66577

**Pull request merged**